ethcore/src/state/mod.rs

```diff
@@ -539,8 +542,55 @@ impl<B: Backend> State<B> {
 			|a| a.as_ref().and_then(|account| account.storage_root().cloned()))
 	}
 
-	/// Mutate storage of account `address` so that it is `value` for `key`.
-	pub fn storage_at(&self, address: &Address, key: &H256) -> TrieResult<H256> {
+	/// Get the value of storage at a specific checkpoint.
+	pub fn checkpoint_storage_at(&self, checkpoint_index: usize, address: &Address, key: &H256) -> TrieResult<Option<H256>> {
+		let checkpoints = self.checkpoints.borrow();
+		let checkpoints_len = checkpoints.len();
+		if let Some(ref checkpoint) = checkpoints.get(checkpoint_index) {
+			if let Some(entry) = checkpoint.get(address) {
+				match entry {
+					Some(entry) => {
+						match entry.account {
+							Some(ref account) => {
+								if let Some(value) = account.cached_storage_at(key) {
+									Ok(Some(value))
+								} else {
+									// This key has checkpoint entry, but not in the entry's cache.
+									// So it's not modified at all.
+									Ok(Some(self.original_storage_at(address, key)?))
+								}
+							},
+							None => {
+								// The account didn't exist at that point. Return empty value.
+								Ok(Some(H256::new()))
+							},
+						}
+					},
+					None => {
+						// The value was not cached at that checkpoint, meaning it was not modified.
+						Ok(Some(self.original_storage_at(address, key)?))
+					},
+				}
+			} else {
+				// This key does not have a checkpoint entry.
+				if checkpoint_index >= checkpoints_len - 1 {
+					Ok(Some(self.storage_at(address, key)?))
+				} else {
+					self.checkpoint_storage_at(checkpoint_index + 1, address, key)
```

Comment:
@tomusdrw It will at most only be 1024 depth (where we only push several small stack items), but I think this is something we should fix for current state mod implementation.

Reply:
Consider the following checkpoint sequences:

```
orig -> c1 -> set 0 -> c2 -> c3 -> set 1
```

When calling `set 0`, `orig` would be swapped to `c1`'s entry. When calling `set 1`, `set 0` would be swapped to `c3`'s entry. So after this, we have:

```
c1: orig
c2: (empty)
c3: set 0
```

If we want to get `c2`'s value (which should be `set 0`), we go to `c2`'s entry, finding that it doesn't even contain the entry, so we need to go downward to `c3` and find `set 0`.

Reply:
You're right, I misunderstood the direction in which checkpoints work.